### PR TITLE
Persist Imgix cache to MMKV

### PR DESCRIPTION
**Pull request by tchayen**
Fixes RNBW-3326

## What changed (plus any additional context for devs)

For a wallet which loads a total of 1k+ images on start, for additional 4ms at startup we save 100ms in runtime. Tested on middle class Android device. User experience difference is IMO much bigger as app feels less laggy and images seem to load faster.

## PoW (screenshots / screen recordings)

Check this thread:
https://rainbowhaus.slack.com/archives/C03B2U2Q4SG/p1652277998457179

## Dev checklist for QA: what to test

Any missing NFT or coin icon images are a sign of problems.

1. Both Android and iOS need checking.
2. Check fresh install, to see how app behaves without cache.
3. Restart app, to see how loading from cache works (is it faster than before? do all images show up?).
4. Minimize the app and wait a bit of time. Then go back and see if anything breaks.

## Final checklist

- [x] Assigned individual reviewers?
- [x] Added labels?
- [x] Added e2e tests? if not please specify why
- [x] If you added new files, did you update the CODEOWNERS file?

### CODE REVIEW

File: src/handlers/imgix.ts [hunk 1]
```diff
@@ -39,8 +50,64 @@ const staticImgixClient = shouldCreateImgixClient();
...
+> => {
+  try {
+    const cache = new LRUCache<string, string>(capacity);
+    const keys = imgixCacheStorage.getString(ATTRIBUTES.KEYS)?.split(',') ?? [];
+    const values =
+      imgixCacheStorage.getString(ATTRIBUTES.VALUES)?.split(',') ?? [];
+
+    for (let i = 0; i < keys.length; i++) {
+      cache.set(keys[i], values[i]);
+    }
```
Comment (terrysahaidak) [inline src/handlers/imgix.ts:69]: I would measure if having it as object is faster in general

```
const entires  = cache.entries();
mmkv.set('cache', JSON.stringify(entries));

// later on

const cached = mmkv.getString('cache');
const cachedEntries = JSON.parse(cached);

cache.from(Object.fromEtnries(entries));


```

File: src/handlers/imgix.ts [hunk 2]
```diff
@@ -39,8 +50,64 @@ const staticImgixClient = shouldCreateImgixClient();
...
+
+    imgixCacheStorage.set(ATTRIBUTES.KEYS, keys.join(','));
+    imgixCacheStorage.set(ATTRIBUTES.VALUES, values.join(','));
+  } catch (error) {
+    logger.error(`Failed to persist IMGIX cache: ${error}`);
+  }
+};
+
+const timeout = 30_000;
```
Comment (brunobar79) [inline src/handlers/imgix.ts:98]: this should be illegal!!! 😂

**Review by brunobar79 — APPROVED** [review]
good stuff! 🚀

**Comment by brunobar79** [comment]
Looks like e2e failed @tchayen

File: src/handlers/imgix.ts (hunk 2 above)
Comment (tchayen) [reply src/handlers/imgix.ts:98]: This is a part where least thinking went into. I just assumed that I will dump it to memory every 30s, but this potentially has many problems.

Will it break when app is in background? Is it maybe too often?

@terrysahaidak I'd also love to hear what you think about this part.

File: src/handlers/imgix.ts (hunk 1 above)
Comment (tchayen) [reply src/handlers/imgix.ts:69]: `cache.entries()` returns an iterator and when passed to `JSON.stringify` it turns into `{"done":false}`, so I changed it to reuse existing code for iterating over it.

Saving:

```tsx
const entries = [];

const iterator = staticSignatureLRU.entries();
for (let i = 0; i < staticSignatureLRU.size; i++) {
  const [key, value] = iterator.next().value;
  entries.push([key, value]);
}

imgixCacheStorage.set('cache', JSON.stringify(entries));
```

Reading:
```tsx
const cacheString = imgixCacheStorage.getString('cache') || '';
const entries = JSON.parse(cacheString);
return LRUCache.from(Object.fromEntries(entries), capacity);
```

This is consistently 1.5-2x slower than reading keys and values separately.

**Review by jkadamczyk — APPROVED** [review]
Looks good to me!

File: src/handlers/imgix.ts [hunk 3]
```diff
@@ -39,8 +50,64 @@ const staticImgixClient = shouldCreateImgixClient();
...
+    for (let i = 0; i < staticSignatureLRU.size; i++) {
+      const [key, value] = iterator.next().value;
+      keys.push(key);
+      values.push(value);
+    }
+
+    imgixCacheStorage.set(ATTRIBUTES.KEYS, keys.join(','));
+    imgixCacheStorage.set(ATTRIBUTES.VALUES, values.join(','));
+  } catch (error) {
+    logger.error(`Failed to persist IMGIX cache: ${error}`);
```
[the remark this answers is not in the document]
Comment (osdnk) [reply src/handlers/imgix.ts:94]: ImgIX

File: src/handlers/imgix.ts [hunk 4]
```diff
@@ -39,8 +50,64 @@ const staticImgixClient = shouldCreateImgixClient();
...
+  try {
+    const keys = [];
+    const values = [];
+
+    const iterator = staticSignatureLRU.entries();
+    for (let i = 0; i < staticSignatureLRU.size; i++) {
+      const [key, value] = iterator.next().value;
+      keys.push(key);
+      values.push(value);
+    }
```
Comment (osdnk) [inline src/handlers/imgix.ts:89]:
```suggestion
    for (let [key, value] of staticSignatureLRU.entries()) {
      keys.push(key);
      values.push(value);
    }
```

File: src/handlers/imgix.ts (hunk 2 above)
Comment (osdnk) [reply src/handlers/imgix.ts:98]: Can you use just `30000`?

File: src/handlers/imgix.ts [hunk 5]
```diff
@@ -39,8 +50,64 @@ const staticImgixClient = shouldCreateImgixClient();
...
+
+const timeout = 30_000;
+const loopSaving = async () => {
+  await saveToMemory();
+  setTimeout(loopSaving, timeout);
+};
+
+setTimeout(() => {
+  loopSaving();
+}, timeout);
```
Comment (osdnk) [inline src/handlers/imgix.ts:106]:
```suggestion
setInterval(() => {
  saveToMemory();
}, timeout);
```

Is there a reason why you made this that complicated is this takes 8ms according to what you wrote?

**Review by osdnk — APPROVED** [review]

File: src/handlers/imgix.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tchayen) [reply src/handlers/imgix.ts:94]: Noooo, their website never uses this spelling.

**Review by terrysahaidak — APPROVED** [review]

**Review by skylarbarrera — APPROVED** [review]
🌞

**Comment by jinchung** [comment]
@tchayen linting errors

**Review by BrodyHughes — APPROVED** [review]
Some small performance issues on complete uninstall / reinstall (likely unrelated directly to this PR) but nothing app breaking I noticed!

**Pull request merged**